### DISCUSSION TITLE
Allow terminal input pane to be pinned to top

### DISCUSSION THREAD

**Discussion by erickvasm** [ghostty-org/ghostty#3444] [Feature Requests, Ideas]
# Feature Request: Pin Input Pane to Top in Ghostty Terminal

![Image](https://github.com/user-attachments/assets/f9e269a5-5e4a-4ef3-846e-e80c079f4917)

## Summary
Introduce an option to pin the terminal input pane to the top of the window for enhanced usability.

## Problem
Currently, users must scroll to access the input pane, which disrupts workflows during intensive terminal sessions. A fixed input pane would streamline command entry, eliminating the need to scroll. Adding this to Ghostty Terminal would improve the experience for its users.

## Performance
Minimal performance impact expected, as this is primarily a UI enhancement.

### General Issue
This feature addresses a broader problem of inefficient navigation in terminal applications, enhancing usability and workflow efficiency.

## Reasons for Implementing the Improvement

- **Eases Command Writing**: Eliminates the need to scroll to access the input pane, enabling seamless command entry.
- **Improves Visibility**: Keeps the input pane fixed, ensuring uninterrupted view of command results.
- **Reduces Visual Fatigue**: Less scrolling leads to reduced strain and increased comfort.
- **Boosts Productivity**: Minimizes distractions caused by navigation, allowing users to focus on their work.

## References

- [Warp Terminal Feature Request](https://github.com/warpdotdev/warp/issues/22)
- [iTerm2 Issue Discussion](https://gitlab.com/gnachman/iterm2/-/issues/11595)

**Comment by chndler** [comment]
+1, personally would want to pin to bottom along with some customizable padding/separator to visually distinguish between the input and previous output

**Comment by JeroenKnoops** [comment]
I've used to pinned to top feature in Warp for several weeks and I really really really liked it. I'm working a lot on a laptop in the terminal, and I've noticed the ergonomical changes. Not constantly looking down to the bottom part of my screen.

So if ghostty can have this option as well, I'm very happy.

**Comment by 00-kat** [comment]
FYI Ghostty has no "input pane" currently.  That thing you can input to is your shell.

That is to say, this feature request *actually* requests to "Add an input pane similar to Warp".

**Comment by pluiedev** [comment]
Related to #3603

**Reply by 101v** [reply]
On a vertical screen looking down hurts the neck. It would really be helpful to have a feature similar to warp where the input pane remains at the top.

**Reply by Leomedina** [reply]
+1 to this

**Comment by mmuellersk** [comment]
I would be very interested in contributing to this feature.

My preferred terminal UX is a permanently top-anchored shell prompt, with command output and history displayed below it, similar to Warp's Reverse mode.

I would like to explore implementing this in Ghostty while preserving standard shell functionality, including autocomplete, command history, ANSI formatting, and interactive applications.

Would the maintainers be open to discussing an architectural approach for this feature, potentially starting with a Linux implementation?

I'd be happy to investigate the codebase and work on an initial prototype if this aligns with the project's direction.

**Reply by binaryben** [reply]
It could be worth incorporating the scope from [Text Blocks or Command Blocks](https://github.com/ghostty-org/ghostty/discussions/7607) (#7607), particularly to support reverse mode. That workflow is exactly what I'm most keen to see implemented as well. Any reason you're anticipating it would be Linux specific?

**Reply by 00-kat** [reply]
> Any reason you're anticipating it would be Linux specific?

They say *starting with*, so I assume they're more familiar with GTK or don't have macOS to develop for it.
